diff --git a/config/install/gin.settings.yml b/config/install/gin.settings.yml
index cb3e5c3c837acfb9070fc838ff99cda66c50c2d1..c05317dc0e01ac58c84ac771eb80912fd9553818 100644
--- a/config/install/gin.settings.yml
+++ b/config/install/gin.settings.yml
@@ -5,6 +5,7 @@ classic_toolbar: 'vertical'
 logo:
   use_default: true
 high_contrast_mode: false
+show_description_toggle: true
 show_user_theme_settings: false
 third_party_settings:
   shortcut:
diff --git a/config/schema/gin.schema.yml b/config/schema/gin.schema.yml
index d03607c9f3568031389c79bbc3a3f44934c5f43b..8ac5bab840137ae4cba14161a011552c5c0beeb5 100644
--- a/config/schema/gin.schema.yml
+++ b/config/schema/gin.schema.yml
@@ -24,6 +24,9 @@ gin.settings:
     focus_color:
       type: string
       label: 'The focus color'
+    show_description_toggle:
+      type: boolean
+      label: 'Display help icon to toggle form description'
     show_user_theme_settings:
       type: boolean
       label: 'Let user override theme'
diff --git a/dist/css/base/gin.css b/dist/css/base/gin.css
index 1f694c4e95229cc0715becbbc74e141690a676cb..f9db8cbcef20597ce53b7ee392b6af23160442e4 100644
--- a/dist/css/base/gin.css
+++ b/dist/css/base/gin.css
@@ -2287,6 +2287,12 @@ fieldset:not(.fieldgroup) > .fieldset-wrapper {
   margin-top: 20px;
 }
 
+/* Align fieldset label positioning to regular form item label, see claro form.css. */
+
+.fieldset__label {
+  margin-top: .25rem;
+}
+
 .fieldset__label,
 .fieldset__label--group,
 .form-item__label {
diff --git a/dist/css/components/description_toggle.css b/dist/css/components/description_toggle.css
new file mode 100644
index 0000000000000000000000000000000000000000..cd3586fb2e16bea60907b41576f00f2c27a3e9e5
--- /dev/null
+++ b/dist/css/components/description_toggle.css
@@ -0,0 +1,93 @@
+/* Base wrapper */
+
+.help-icon {
+  display: flex;
+}
+
+/* Size of svg + border added by claro. */
+
+.help-icon__description-toggle {
+  all: unset;
+  flex-shrink: 0;
+  margin: .3rem 1rem 0 .3rem;
+  height: 16px;
+  width: 16px;
+  border-radius: 50%;
+}
+
+.help-icon__description-toggle:before {
+  background-color: var(--colorGinIcons);
+  content: "";
+  display: block;
+  width: 100%;
+  height: 100%;
+  -webkit-mask-image: url("../../media/sprite.svg#help-view");
+          mask-image: url("../../media/sprite.svg#help-view");
+  -webkit-mask-repeat: no-repeat;
+          mask-repeat: no-repeat;
+  -webkit-mask-position: center center;
+          mask-position: center center;
+  -webkit-mask-size: 100% 100%;
+          mask-size: 100% 100%;
+}
+
+.help-icon__description-toggle:hover:before {
+  background-color: var(--colorGinPrimary);
+}
+
+.form-type--checkbox > .help-icon__element-has-description,
+.form-type--radio > .help-icon__element-has-description {
+  display: flex;
+  align-items: normal;
+}
+
+.form-type--checkbox > .help-icon__element-has-description > label,
+.form-type--radio > .help-icon__element-has-description > label {
+  margin-left: 6px;
+}
+
+.form-type--checkbox .checkbox-toggle + .help-icon,
+.form-type--radio .checkbox-toggle + .help-icon {
+  padding-left: 4px;
+}
+
+/*
+ * 1. Subtract description toggle width from input fields.
+ */
+
+.help-icon__element-has-description input:not(.form-checkbox):not(.form-autocomplete):not(.form-file),
+.help-icon__element-has-description select {
+  max-width: calc(100% - 34px);
+  /* 1 */
+}
+
+/* Float toggle in detail summary. */
+
+details.help-icon__description-container .help-icon__description-toggle {
+  float: right;
+}
+
+/* Fieldgroup adjustments */
+
+fieldset:not(.fieldgroup) .help-icon > .fieldset__legend .fieldset__label {
+  padding-right: 0;
+}
+
+fieldset.media-library-widget .help-icon > .help-icon__description-toggle {
+  margin-top: 1.3rem;
+}
+
+/* System Modules Search fields */
+
+.system-modules .table-filter .form-type--search,
+.system-modules-uninstall .table-filter .form-type--search {
+  position: relative;
+}
+
+.system-modules .table-filter .form-type--search .help-icon__description-toggle,
+.system-modules-uninstall .table-filter .form-type--search .help-icon__description-toggle {
+  position: absolute;
+  right: -8px;
+  top: 16px;
+}
+
diff --git a/dist/css/theme/dialog.css b/dist/css/theme/dialog.css
index d801e27665ad8397436cd064e677fa1c62f34bf6..cdbe4d63cc426a63bd22eb03b2bface2853650e7 100644
--- a/dist/css/theme/dialog.css
+++ b/dist/css/theme/dialog.css
@@ -591,6 +591,10 @@
   margin-top: 20px;
 }
 
+.ui-dialog .fieldset__label {
+  margin-top: .25rem;
+}
+
 .ui-dialog .fieldset__label,
 .ui-dialog .fieldset__label--group,
 .ui-dialog .form-item__label {
diff --git a/dist/js/description_toggle.js b/dist/js/description_toggle.js
new file mode 100644
index 0000000000000000000000000000000000000000..13fa3cacc6d7a0571494b3e90926c05e659990a9
--- /dev/null
+++ b/dist/js/description_toggle.js
@@ -0,0 +1,14 @@
+(Drupal => {
+  Drupal.behaviors.formDescriptionToggle = {
+    attach: function(context) {
+      context.querySelectorAll(".help-icon__description-toggle").forEach((elem => {
+        elem.dataset.formDescriptionToggleAttached || (elem.dataset.formDescriptionToggleAttached = !0, 
+        elem.addEventListener("click", (event => {
+          event.preventDefault(), event.currentTarget.focus(), event.currentTarget.closest(".help-icon__description-container").querySelectorAll(".claro-details__description, .fieldset__description, .form-item__description").forEach((description => {
+            description.classList.toggle("visually-hidden");
+          }));
+        })));
+      }));
+    }
+  };
+})(Drupal);
\ No newline at end of file
diff --git a/gin.libraries.yml b/gin.libraries.yml
index d4f6905e371c5948e688c3a6050bb5e99ab8e31a..d94db328b475493f46911106793a00128161da54 100644
--- a/gin.libraries.yml
+++ b/gin.libraries.yml
@@ -89,6 +89,13 @@ gin_dialog:
     - gin/gin_base
     - gin/gin_accent
 
+gin_description_toggle:
+  css:
+    theme:
+      dist/css/components/description_toggle.css: { minified: false }
+  js:
+    dist/js/description_toggle.js: { minified: false }
+
 gin_messages:
   version: VERSION
   js:
diff --git a/includes/form.theme b/includes/form.theme
index 9e5446b81593be5f391c65692434b673fa7c2083..54d18b4b4b740891147db8a3382e30a9a6996b3e 100644
--- a/includes/form.theme
+++ b/includes/form.theme
@@ -6,6 +6,7 @@
  */
 
 use Drupal\Core\Form\FormStateInterface;
+use Drupal\gin\GinPreRender;
 use Drupal\gin\GinContentFormHelper;
 use Drupal\gin\GinSettings;
 
@@ -119,6 +120,57 @@ function gin_theme_suggestions_input_alter(array &$suggestions, array $variables
   }
 }
 
+/**
+ * Implements template_preprocess_HOOK() for form_element.
+ */
+function gin_preprocess_form_element(&$variables) {
+  _gin_add_description_toggle($variables);
+}
+
+/**
+ * Implements template_preprocess_HOOK() for datetime_wrapper.
+ */
+function gin_preprocess_datetime_wrapper(&$variables) {
+  _gin_add_description_toggle($variables);}
+
+/**
+ * Implements template_preprocess_HOOK() for details.
+ */
+function gin_preprocess_details(&$variables) {
+  _gin_add_description_toggle($variables);
+}
+
+/**
+ * Implements template_preprocess_HOOK() for fieldset.
+ */
+function gin_preprocess_fieldset(&$variables) {
+  _gin_add_description_toggle($variables);
+}
+
+/**
+ * Implements hook_element_info_alter().
+ */
+function gin_element_info_alter(&$info) {
+  if (array_key_exists('text_format', $info)) {
+    $info['text_format']['#pre_render'][] = [
+      GinPreRender::class,
+      'textFormat',
+    ];
+  }
+}
+
+/**
+ * Implements template_preprocess_HOOK() for text_format_wrapper.
+ */
+function gin_preprocess_text_format_wrapper(&$variables) {
+  /** @var \Drupal\gin\GinSettings $settings */
+  $settings = \Drupal::classResolver(GinSettings::class);
+  if ($settings->get('show_description_toggle') && !empty($variables['description'])) {
+    $variables['description_display'] = 'invisible';
+    $variables['description_toggle'] = TRUE;
+  }
+}
+
 /**
  * Implements template_preprocess_inline_entity_form_entity_table() for forms.
  */
diff --git a/includes/helper.theme b/includes/helper.theme
index 7ef3c09b8f511e25f5d1538525656c31ec1f945e..293c2cf5ca043bb5261fa62418177bc1df3c4c66 100644
--- a/includes/helper.theme
+++ b/includes/helper.theme
@@ -56,6 +56,7 @@ function _gin_user_form_submit(&$form, FormStateInterface $form_state) {
       'preset_focus_color' => $form_state->getValue('preset_focus_color'),
       'focus_color' => $form_state->getValue('focus_color'),
       'high_contrast_mode' => (bool) $form_state->getValue('high_contrast_mode'),
+      'show_description_toggle' => $form_state->getValue('show_description_toggle'),
     ];
     $settings->setAll($user_settings, $account);
   }
@@ -98,3 +99,75 @@ function _gin_is_active() {
 
   return $gin_activated;
 }
+
+/**
+ * Check if were on a content edit form.
+ */
+function _gin_is_content_form($form = NULL, $form_state = NULL, $form_id = NULL) {
+  $is_content_form = FALSE;
+
+  // Get route name.
+  $route_name = \Drupal::routeMatch()->getRouteName();
+
+  // Routes to include.
+  $route_names = [
+    'node.add',
+    'entity.node.content_translation_add',
+    'entity.node.content_translation_edit',
+    'quick_node_clone.node.quick_clone',
+    'entity.node.edit_form',
+  ];
+
+  if (
+    in_array($route_name, $route_names, TRUE) ||
+    ($form_state && ($form_state->getBuildInfo()['base_form_id'] ?? NULL) === 'node_form') ||
+    ($route_name === 'entity.group_content.create_form' && strpos($form_id, 'group_node') === FALSE)
+  ) {
+    $is_content_form = TRUE;
+  }
+
+  // Forms to exclude.
+  // If media library widget, don't use new content edit form.
+  // gin_preprocess_html is not triggered here, so checking
+  // the form id is enough.
+  $form_ids_to_ignore = [
+    'media_library_add_form_',
+    'views_form_media_library_widget_',
+    'views_exposed_form',
+    'date_recur_modular_sierra_occurrences_modal',
+  ];
+
+  foreach ($form_ids_to_ignore as $form_id_to_ignore) {
+    if ($form_id && strpos($form_id, $form_id_to_ignore) !== FALSE) {
+      $is_content_form = FALSE;
+    }
+  }
+
+  return $is_content_form;
+}
+
+/**
+ * @param $variables
+ */
+function _gin_add_description_toggle(&$variables) {
+  if (_gin_is_description_toggle_enabled()){
+    if (!empty($variables['description'])) {
+      $variables['description_display'] = 'invisible';
+      $variables['description_toggle'] = TRUE;
+    }
+    // Add toggle for text_format, description is in wrapper.
+    elseif (!empty($variables['element']['#description_toggle'])) {
+      $variables['description_toggle'] = TRUE;
+    }
+  }
+}
+
+/**
+ * @return array|bool|mixed|string|null
+ */
+function _gin_is_description_toggle_enabled() {
+  /** @var \Drupal\gin\GinSettings $settings */
+  $settings = \Drupal::classResolver(GinSettings::class);
+  return $settings->get('show_description_toggle') && _gin_is_content_form();
+}
+
diff --git a/js/description_toggle.js b/js/description_toggle.js
new file mode 100644
index 0000000000000000000000000000000000000000..32e6e49c4c65756b5ec331aec2a113f5596ab302
--- /dev/null
+++ b/js/description_toggle.js
@@ -0,0 +1,28 @@
+((Drupal) => {
+  Drupal.behaviors.formDescriptionToggle = {
+    attach: function attach(context) {
+      context
+        .querySelectorAll('.help-icon__description-toggle')
+        .forEach((elem) => {
+          if (elem.dataset.formDescriptionToggleAttached) {
+            return;
+          }
+          elem.dataset.formDescriptionToggleAttached = true;
+
+          elem.addEventListener('click', (event) => {
+            const toggleFunction = (description) => {
+              description.classList.toggle('visually-hidden');
+            };
+            event.preventDefault();
+            event.currentTarget.focus(); // firefox button focus issue
+            event.currentTarget
+              .closest('.help-icon__description-container')
+              .querySelectorAll(
+                '.claro-details__description, .fieldset__description, .form-item__description',
+              )
+              .forEach(toggleFunction);
+          });
+        });
+    },
+  };
+})(Drupal);
diff --git a/src/GinPreRender.php b/src/GinPreRender.php
new file mode 100644
index 0000000000000000000000000000000000000000..775ddd65607866219b848426a070b04f163316a9
--- /dev/null
+++ b/src/GinPreRender.php
@@ -0,0 +1,41 @@
+<?php
+
+namespace Drupal\gin;
+
+use Drupal\Core\Security\TrustedCallbackInterface;
+
+/**
+ * Implements trusted prerender callbacks for the Gin theme.
+ *
+ * @internal
+ */
+class GinPreRender implements TrustedCallbackInterface {
+
+  /**
+   * Prepare description toggle for output in template.
+   */
+  public static function textFormat($element) {
+
+    if (_gin_is_description_toggle_enabled() && !empty($element['#description'])) {
+      if ($element['#type'] === 'text_format') {
+        $element['value']['#description_toggle'] = TRUE;
+      }
+      else {
+        $element['#description_toggle'] = TRUE;
+        $element['#description_display'] = 'invisible';
+      }
+
+    }
+     return $element;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function trustedCallbacks() {
+    return [
+      'textFormat',
+    ];
+  }
+
+}
diff --git a/src/GinSettings.php b/src/GinSettings.php
index 9b4a1f79ecbd2d4d577758dea9532a41c16f1952..21686cb053768399d002b949352987ef8e4b086b 100644
--- a/src/GinSettings.php
+++ b/src/GinSettings.php
@@ -315,7 +315,7 @@ class GinSettings implements ContainerInjectionInterface {
       '#title' => $this->t('Custom Accent color'),
       '#description' => $this->t('Use with caution, values should meet a11y criteria.'),
       '#states' => [
-      // Show if met.
+        // Show if met.
         'visible' => [
           ':input[name="preset_accent_color"]' => ['value' => 'custom'],
         ],
@@ -401,6 +401,14 @@ class GinSettings implements ContainerInjectionInterface {
       ],
     ];
 
+    $form['show_description_toggle'] = [
+      '#type' => 'checkbox',
+      '#title' => t('Enable form description toggle.'),
+      '#description' => t('Show a help icon to show/hide form descriptions on content forms.'),
+      '#default_value' => $account ? $this->get('show_description_toggle', $account) : $this->getDefault('show_description_toggle'),
+      '#weight' => 0,
+    ];
+
     // High contrast mode.
     $form['high_contrast_mode'] = [
       '#type' => 'checkbox',
diff --git a/styles/base/_form.scss b/styles/base/_form.scss
index 5fb84318ae3c2f4d1e7510de5d67802c00265e50..cd2e820a24dcce3a3aff75136bc7ca7f89c7cad0 100644
--- a/styles/base/_form.scss
+++ b/styles/base/_form.scss
@@ -359,6 +359,11 @@ fieldset:not(.fieldgroup) {
   }
 }
 
+/* Align fieldset label positioning to regular form item label, see claro form.css. */
+.fieldset__label {
+  margin-top: 0.25rem;
+}
+
 .fieldset__label,
 .fieldset__label--group,
 .form-item__label {
diff --git a/styles/components/description_toggle.scss b/styles/components/description_toggle.scss
new file mode 100644
index 0000000000000000000000000000000000000000..3abb3c624dd0c48351107c9d8bbdfc9ddfa21a7e
--- /dev/null
+++ b/styles/components/description_toggle.scss
@@ -0,0 +1,90 @@
+/* Base wrapper */
+.help-icon {
+  display: flex;
+}
+
+/* Size of svg + border added by claro. */
+.help-icon__description-toggle {
+  all: unset;
+  flex-shrink: 0;
+  margin: .3rem 1rem 0 .3rem;
+  height: 16px;
+  width: 16px;
+  border-radius: 50%;
+
+  &:before {
+    background-color: var(--colorGinIcons);
+    content: "";
+    display: block;
+    width: 100%;
+    height: 100%;
+    mask-image: icon('help');
+    mask-repeat: no-repeat;
+    mask-position: center center;
+    mask-size: 100% 100%;
+  }
+
+  &:hover:before {
+    background-color: var(--colorGinPrimary);
+  }
+}
+
+.form-type--checkbox,
+.form-type--radio {
+  > .help-icon__element-has-description {
+    display: flex;
+    align-items: normal;
+
+    > label {
+      margin-left: 6px;
+    }
+  }
+
+  .checkbox-toggle + .help-icon {
+    padding-left: 4px;
+  }
+}
+
+/*
+ * 1. Subtract description toggle width from input fields.
+ */
+.help-icon__element-has-description {
+  input:not(.form-checkbox):not(.form-autocomplete):not(.form-file),
+  select {
+    max-width: calc(100% - 34px); /* 1 */
+  }
+}
+
+/* Float toggle in detail summary. */
+details.help-icon__description-container .help-icon__description-toggle {
+  float: right;
+}
+
+/* Fieldgroup adjustments */
+fieldset:not(.fieldgroup) {
+  .help-icon {
+    > .fieldset__legend .fieldset__label {
+      padding-right: 0;
+    }
+  }
+}
+
+fieldset.media-library-widget {
+  .help-icon {
+    > .help-icon__description-toggle {
+      margin-top: 1.3rem;
+    }
+  }
+}
+
+/* System Modules Search fields */
+.system-modules .table-filter .form-type--search,
+.system-modules-uninstall .table-filter .form-type--search {
+  position: relative;
+
+  .help-icon__description-toggle {
+    position: absolute;
+    right: -8px;
+    top: 16px;
+  }
+}
diff --git a/templates/form/datetime-wrapper.html.twig b/templates/form/datetime-wrapper.html.twig
new file mode 100644
index 0000000000000000000000000000000000000000..f0faa05e1b8c0ff04fd8e339f0033c15ba6661a2
--- /dev/null
+++ b/templates/form/datetime-wrapper.html.twig
@@ -0,0 +1,55 @@
+{#
+/**
+ * @file
+ * Theme override of a datetime form wrapper.
+ *
+ * @todo Refactor when https://www.drupal.org/node/3010558 is fixed.
+ *
+ * @see template_preprocess_form_element()
+ */
+#}
+{%
+  set classes = [
+    'form-item',
+    'form-datetime-wrapper',
+    description_toggle ? 'help-icon__description-container',
+  ]
+  %}
+{%
+  set title_classes = [
+    'form-item__label',
+    required ? 'js-form-required',
+    required ? 'form-required',
+    errors ? 'has-error',
+  ]
+%}
+{%
+  set description_classes = [
+    'form-item__description',
+    description_display == 'invisible' ? 'visually-hidden',
+  ]
+%}
+<div {{ attributes.addClass(classes) }}>
+  {% if description and description_toggle %}
+  <div class="help-icon">
+  {% endif %}
+  {% if title %}
+  <h4{{ title_attributes.addClass(title_classes) }}>{{ title }}</h4>
+  {% endif %}
+  {% if description and description_toggle %}
+    {{ attach_library('gin/gin_description_toggle') }}
+    <button class="help-icon__description-toggle"></button>
+  </div>
+  {% endif %}
+  {{ content }}
+{% if errors %}
+  <div class="form-item__error-message">
+    {{ errors }}
+  </div>
+{% endif %}
+{% if description %}
+  <div{{ description_attributes.addClass(description_classes) }}>
+    {{ description }}
+  </div>
+{% endif %}
+</div>
diff --git a/templates/form/details.html.twig b/templates/form/details.html.twig
index 78534960770dedb127040d204e611222270995a2..2c501bfb4f2bfe8641d0a08581e6c2cface19762 100644
--- a/templates/form/details.html.twig
+++ b/templates/form/details.html.twig
@@ -29,6 +29,7 @@
     'claro-details',
     accordion ? 'claro-details--accordion',
     accordion_item ? 'claro-details--accordion-item',
+    description_toggle ? 'help-icon__description-container',
   ]
 %}
 {%
@@ -46,6 +47,13 @@
     accordion_item ? 'claro-details__content--accordion-item',
   ]
 %}
+{%
+  set description_classes = [
+  'claro-details__description',
+  disabled ? 'is-disabled',
+  description_display == 'invisible' ? 'visually-hidden',
+]
+%}
 <details{{ attributes.addClass(classes) }}>
   {%- if title -%}
     {%
@@ -60,6 +68,10 @@
     <summary{{ summary_attributes.addClass(summary_classes) }}>
       {{- title -}}
       <span class="required-mark"></span>
+      {% if description and description_toggle %}
+        {{ attach_library('gin/gin_description_toggle') }}
+        <button class="help-icon__description-toggle"></button>
+      {% endif %}
     </summary>
   {%- endif -%}
   <div{{ content_attributes.addClass(content_wrapper_classes) }}>
@@ -73,7 +85,7 @@
         </div>
       {% endif %}
       {%- if description -%}
-        <div class="claro-details__description{{ disabled ? ' is-disabled' }}">{{ description }}</div>
+        <div{{ create_attribute().addClass(description_classes) }}>{{ description }}</div>
       {%- endif -%}
       {%- if children -%}
         {{ children }}
diff --git a/templates/form/fieldset.html.twig b/templates/form/fieldset.html.twig
new file mode 100644
index 0000000000000000000000000000000000000000..8a02c80fd13fb60b04fd88ebb05e135942165325
--- /dev/null
+++ b/templates/form/fieldset.html.twig
@@ -0,0 +1,89 @@
+{#
+/**
+ * @file
+ * Theme override for a fieldset element and its children.
+ *
+ * @see template_preprocess_fieldset()
+ * @see claro_preprocess_fieldset()
+ */
+#}
+{%
+  set classes = [
+  'fieldset',
+  attributes.hasClass('fieldgroup') ? 'fieldset--group',
+  'js-form-item',
+  'form-item',
+  'js-form-wrapper',
+  'form-wrapper',
+  description_toggle ? 'help-icon__description-container',
+]
+%}
+{%
+  set wrapper_classes = [
+  'fieldset__wrapper',
+  attributes.hasClass('fieldgroup') ? 'fieldset__wrapper--group',
+]
+%}
+{%
+  set legend_span_classes = [
+  'fieldset__label',
+  attributes.hasClass('fieldgroup') ? 'fieldset__label--group',
+  required ? 'js-form-required',
+  required ? 'form-required',
+]
+%}
+{%
+  set legend_classes = [
+  'fieldset__legend',
+  attributes.hasClass('fieldgroup') and not attributes.hasClass('form-composite') ? 'fieldset__legend--group',
+  attributes.hasClass('form-composite') ? 'fieldset__legend--composite',
+  title_display == 'invisible' ? 'fieldset__legend--invisible' : 'fieldset__legend--visible',
+]
+%}
+{%
+  set description_classes = [
+  'fieldset__description',
+  description_display == 'invisible' ? 'visually-hidden',
+]
+%}
+<fieldset{{ attributes.addClass(classes) }}>
+  {% if description.content and description_toggle %}
+    <div class="help-icon">
+  {% endif %}
+  {#  Always wrap fieldset legends in a <span> for CSS positioning. #}
+  {% if legend.title %}
+    <legend{{ legend.attributes.addClass(legend_classes) }}>
+      <span{{ legend_span.attributes.addClass(legend_span_classes) }}>{{ legend.title }}</span>
+    </legend>
+  {% endif %}
+  {% if description.content and description_toggle %}
+      {{ attach_library('gin/gin_description_toggle') }}
+      <button class="help-icon__description-toggle"></button>
+    </div>
+  {% endif %}
+  <div{{ content_attributes.addClass(wrapper_classes) }}>
+    {% if inline_items %}
+    <div class="container-inline">
+      {% endif %}
+
+      {% if prefix %}
+        <span class="fieldset__prefix">{{ prefix }}</span>
+      {% endif %}
+      {{ children }}
+      {% if suffix %}
+        <span class="fieldset__suffix">{{ suffix }}</span>
+      {% endif %}
+      {% if errors %}
+        <div class="fieldset__error-message">
+          {{ errors }}
+        </div>
+      {% endif %}
+      {% if description.content %}
+        <div{{ description.attributes.addClass(description_classes) }}>{{ description.content }}</div>
+      {% endif %}
+
+      {% if inline_items %}
+    </div>
+    {% endif %}
+  </div>
+</fieldset>
diff --git a/templates/form/form-element.html.twig b/templates/form/form-element.html.twig
new file mode 100644
index 0000000000000000000000000000000000000000..8eae11df5f4d4f936b35ff67121bb5838631e454
--- /dev/null
+++ b/templates/form/form-element.html.twig
@@ -0,0 +1,85 @@
+{#
+/**
+ * @file
+ * Theme override for a form element.
+ *
+ * @see template_preprocess_form_element()
+ */
+#}
+{#
+Most of core-provided js assumes that the CSS class pattern js-form-item-[something] or
+js-form-type-[something] exists on form items. We have to keep them.
+#}
+{%
+  set classes = [
+    'js-form-item',
+    'form-item',
+    'js-form-type-' ~ type|clean_class,
+    'form-type--' ~ type|clean_class,
+    type in ['checkbox', 'radio'] ? 'form-type--boolean',
+    'js-form-item-' ~ name|clean_class,
+    'form-item--' ~ name|clean_class,
+    title_display not in ['after', 'before'] ? 'form-item--no-label',
+    disabled == 'disabled' ? 'form-item--disabled',
+    errors ? 'form-item--error',
+    description.content ? 'help-icon__description-container'
+  ]
+%}
+{%
+  set description_classes = [
+    'form-item__description',
+    description_display == 'invisible' ? 'visually-hidden',
+  ]
+%}
+<div{{ attributes.addClass(classes) }}>
+  {% if label_display in ['before', 'invisible'] %}
+    {% if description_toggle %}
+      <div class="help-icon">
+        {{ label }}
+        {{ attach_library('gin/gin_description_toggle') }}
+        <button class="help-icon__description-toggle"></button>
+      </div>
+    {% else %}
+      {{ label }}
+    {% endif %}
+  {% endif %}
+  {% if description.content and description_toggle %}
+  <div class="help-icon__element-has-description">
+  {% endif %}
+  {% if prefix is not empty %}
+    <span class="form-item__prefix{{disabled == 'disabled' ? ' is-disabled'}}">{{ prefix }}</span>
+  {% endif %}
+  {% if description_display == 'before' and description.content %}
+    <div{{ description.attributes.addClass(description_classes) }}>
+      {{ description.content }}
+    </div>
+  {% endif %}
+  {{ children }}
+  {% if suffix is not empty %}
+    <span class="form-item__suffix{{disabled == 'disabled' ? ' is-disabled'}}">{{ suffix }}</span>
+  {% endif %}
+  {% if label_display == 'after' %}
+    {% if description.content and description_toggle %}
+      <div class="help-icon">
+        {{ label }}
+        {{ attach_library('gin/gin_description_toggle') }}
+        <button class="help-icon__description-toggle"></button>
+      </div>
+    {% else %}
+      {{ label }}
+    {% endif %}
+  {% endif %}
+  {% if description.content and description_toggle %}
+  </div>
+  {% endif %}
+  {% if errors %}
+    <div class="form-item__error-message">
+      {{ errors }}
+    </div>
+  {% endif %}
+  {% if description_display in ['after', 'invisible'] and description.content %}
+    <div{{ description.attributes.addClass(description_classes) }}>
+      {{ description.content }}
+    </div>
+  {% endif %}
+</div>
diff --git a/templates/form/text-format-wrapper.html.twig b/templates/form/text-format-wrapper.html.twig
new file mode 100644
index 0000000000000000000000000000000000000000..5c8711e9121a7ec230116a2b3a8307efae6c035b
--- /dev/null
+++ b/templates/form/text-format-wrapper.html.twig
@@ -0,0 +1,43 @@
+{#
+/**
+ * @file
+ * Theme override for a text format-enabled form element.
+ *
+ * @todo Remove when https://www.drupal.org/node/3016346 and
+ * https://www.drupal.org/node/3016343 are fixed.
+ *
+ * Available variables:
+ * - children: Text format element children.
+ * - description: Text format element description.
+ * - attributes: HTML attributes for the containing element.
+ * - aria_description: Flag for whether or not an ARIA description has been
+ *   added to the description container.
+ * - description_attributes: attributes for the description.
+ *   @see https://www.drupal.org/node/3016343
+ * - disabled: An indicator for whether the associated form element is disabled,
+ *   added by this theme.
+ *   @see https://www.drupal.org/node/3016346
+ *
+ * @see template_preprocess_text_format_wrapper()
+ */
+#}
+{%
+  set classes = [
+  'js-form-item',
+  'form-item',
+  description_toggle ? 'help-icon__description-container',
+]
+%}
+<div{{ attributes.addClass(classes) }}>
+  {{ children }}
+  {% if description %}
+    {%
+      set description_classes = [
+      aria_description ? 'form-item__description',
+      disabled ? 'is-disabled',
+      description_display == 'invisible' ? 'visually-hidden',
+    ]
+    %}
+    <div{{ description_attributes.addClass(description_classes) }}>{{ description }}</div>
+  {% endif %}
+</div>
diff --git a/webpack.config.js b/webpack.config.js
index b6b2b90b0900be12bf77927c35e8064250fdff0e..b0b820687e516c45822f655643fb75bceb32db70 100644
--- a/webpack.config.js
+++ b/webpack.config.js
@@ -16,6 +16,7 @@ module.exports = {
     'messages': ['./js/messages.js'],
     'settings': ['./js/settings.js'],
     'gin_ckeditor': ['./js/gin_ckeditor.js'], // Can't rename as we would be in trouble
+    'description_toggle': ['./js/description_toggle.js'],
     // Base
     'base/gin': ['./styles/gin.scss'],
     // Components
@@ -44,6 +45,7 @@ module.exports = {
     'components/media_library': ['./styles/components/media_library.scss'],
     'components/contextual_links': ['./styles/components/contextual_links.scss'],
     'components/ajax': ['./styles/components/ajax.scss'],
+    'components/description_toggle': ['./styles/components/description_toggle.scss'],
     // Layout
     'layout/toolbar': ['./styles/layout/toolbar.scss'],
     'layout/horizontal_toolbar': ['./styles/layout/horizontal_toolbar.scss'],
